src/Physics/NuclearState/NucleusGenINCL.cxx

```diff
@@ -0,0 +1,336 @@
+  // the removal energy maybe not necessary
+  TVector3 p3 = incl_nucleus->getHitNucleonMomentum();
+  double   hit_nucleon_energy = incl_nucleus->getHitNucleonEnergy();
+  double   w  = incl_nucleus->getRemovalEnergy();
+  //-- update the struck nucleon 4p at the interaction summary and at
+  // the GHEP record
+  p4->SetPx(p3.Px()/1000.);
+  p4->SetPy(p3.Py()/1000.);
+  p4->SetPz(p3.Pz()/1000.);
+  p4->SetE (hit_nucleon_energy/1000.);
+  LOG("NucleusGenINCL", pINFO) << "Momentum";
+  p4->Print();
+
+  nucleon->SetMomentum(*p4);  // update GHEP value
+  nucleon->SetRemovalEnergy(w);  // FIXME this may be not necessary
+
+
+  // Sometimes, for interactions near threshold, Fermi momentum might bring
+  // the neutrino energy in the nucleon rest frame below threshold (for the
+  // selected interaction). In this case mark the event as unphysical and
+  // abort the current thread.
+  const KPhaseSpace & kps = interaction->PhaseSpace();
+  if(!kps.IsAboveThreshold()) {
+     LOG("NucleusGenINCL", pNOTICE)
+                  << "Event below threshold after generating Fermi momentum";
+
+     double Ethr = kps.Threshold();
+     double Ev   = init_state->ProbeE(kRfHitNucRest);
+     LOG("NucleusGenINCL", pNOTICE)
+              << "Ev (@ nucleon rest frame) = " << Ev << ", Ethr = " << Ethr;
+
+     evrec->EventFlags()->SetBitNumber(kBelowThrNRF, true);
+     genie::exceptions::EVGThreadException exception;
+     exception.SetReason("E < Ethr after generating nucleon Fermi momentum");
+     exception.SwitchOnFastForward();
+     throw exception;
+  }
+
+}
+
+void NucleusGenINCL::setTargetNucleusRemnant(GHepRecord * evrec)const{
+// add the remnant nuclear target at the GHEP record
+
+  LOG("NucleusGenINCL", pINFO) << "Adding final state nucleus";
+
+  double Px = 0;
+  double Py = 0;
+  double Pz = 0;
+  double E  = 0;
+
+  GHepParticle * nucleus = evrec->TargetNucleus();
+  int A = nucleus->A();
+  int Z = nucleus->Z();
+
+  int fd = nucleus->FirstDaughter();
+  int ld = nucleus->LastDaughter();
+
+  INCLNucleus *incl_nucleus = INCLNucleus::Instance();
+
+  G4INCL::Nucleus *incl_nuc = incl_nucleus->getNuclues();
+
+
+  for(int id = fd; id <= ld; id++) {
+
+    // compute A,Z for final state nucleus & get its PDG code and its mass
+    GHepParticle * particle = evrec->Particle(id);
+    assert(particle);
+    int  pdgc = particle->Pdg();
+    bool is_p  = pdg::IsProton (pdgc);
+    bool is_n  = pdg::IsNeutron(pdgc);
+
+    if (is_p) Z--;
+    if (is_p || is_n) A--;
+
+    Px += particle->Px();
+    Py += particle->Py();
+    Pz += particle->Pz();
+    E  += particle->E();
+
+  }//daughters
+
+  TParticlePDG * remn = 0;
+  int ipdgc = pdg::IonPdgCode(A, Z);
+  remn = PDGLibrary::Instance()->Find(ipdgc);
+  if(!remn) {
+    LOG("HadronicVtx", pFATAL)
+          << "No particle with [A = " << A << ", Z = " << Z
+                            << ", pdgc = " << ipdgc << "] in PDGLibrary!";
+    assert(remn);
```

Comment:
The earlier comment here about `assert` being a no-op for non-debug code makes this section not do what I think you want it to.   Even if you're hoping that the branch never gets triggered it's best that it does what it is intended to do.